### PR TITLE
Remove INodeVisitor; Write -> Print; switch to init properties

### PR DESCRIPTION
**Pull request by sungam3r**
Helps to refactor ` internal static string Print(this ASTNode node)` in the main repo.

### CODE REVIEW

File: src/GraphQLParser/Annotations/IsExternalInit.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+using System.ComponentModel;
+
+namespace System.Runtime.CompilerServices;
+
+/// <summary>
+/// Reserved to be used by the compiler for tracking metadata.
+/// This class should not be used by developers in source code.
+/// </summary>
+[EditorBrowsable(EditorBrowsableState.Never)]
+internal static class IsExternalInit
```
Comment (sungam3r) [inline src/GraphQLParser/Annotations/IsExternalInit.cs:12]: For init properties support.

**Comment by Shane32** [comment]
Instead of `ToSDL` extension method, how about this:

```cs
public static class SDLWriter
{
    public static async PrintAsync(node, writer, options)
    {
        //code from extension method goes here
    }
}
```

Or:

```cs
public class SDLWriter : INodeVisitor<DefaultWriterContext>
{
    public async ValueTask Print(node, writer, options)
    {
        //code from ToSDL method here
    }

    ValueTask INodeVisitor<DefaultWriteContext>.VisitBlahBlahAsync(....)
    {
    }

    //etc
}
```

In this manner it is still possible to override the behavior of any method of the SDLPrinter.

Another idea: make all `Visit...` methods of the `DefaultNodeVisitor` protected instead of public.

Just thinking of different ideas to clean up `SDLWriter` rather than having all those `Visit...`methods exposed...

**Comment by Shane32** [comment]
I do REALLY like the fact that the entire SDLWriter can be overridden as desired.

**Comment by Shane32** [comment]
If you like the latter idea that I just posted, I'll be happy to write a PR.

File: src/GraphQLParser/Extensions/ASTNodeExtensions.cs [hunk 2]
```diff
@@ -8,6 +10,16 @@ namespace GraphQLParser;
...
 {
+    /// <summary>
+    /// Prints <see cref="ASTNode"/> into the provided <see cref="TextWriter"/> as a SDL document.
+    /// </summary>
+    public static async ValueTask ToSDL(this ASTNode node, TextWriter writer, SDLWriterOptions? options = null)
+    {
+        var visitor = new SDLWriter<DefaultWriteContext>(options ?? new SDLWriterOptions());
+        var context = new DefaultWriteContext(writer);
+        await visitor.Visit(node, context);
+    }
```
[the remark this answers is not in the document]
Comment (Shane32) [reply src/GraphQLParser/Extensions/ASTNodeExtensions.cs]: Whoops I thought I posted this comment a while ago.

**Comment by sungam3r** [comment]
> Just thinking of different ideas to clean up SDLWriter rather than having all those Visit...methods exposed...

They all are intended to be exposed. Visitor pattern, nothing to hide.

**Comment by sungam3r** [comment]
```
public static class SDLWriter
{
    public static async PrintAsync(node, writer, options)
    {
        //code from extension method goes here
    }
}
```

Well, it could be rewritten but for what reason ?

**Comment by sungam3r** [comment]
I would not change public `VisitXXX` to protected.

**Comment by Shane32** [comment]
I don't like creating extension methods that encapsulate classes and complex execution patterns.  It does not seem to be good OOP design.

I also don't like classes that have methods that are obviously used only internally exposed as public when they should be protected.  In this instance, `Visit` provides indirect access to every single method.  There is no reason to have them public.  It just makes it so when someone wants to actually use the class, they have no idea what to do when they see this:

image

It should be a simple class with only `Print` exposed.  It should be a 'printer' class, not a 'visitor'.  It should use a visitor "under the hood".  All other methods should be protected.  Any node can still be printed by calling `Print`.  To change any functionality, just derive from the class.  Nothing changes except it's a very clean design to actually use.  There is no loss in functionality at all.

**Comment by sungam3r** [comment]
>  All other methods should be protected

public **interface** INodeVisitor<TContext>

While I do not understand the scale of all changes and I do not see a sufficient reason to do it except the judgment that too many methods can be seen.

**Comment by Shane32** [comment]
```cs
/// <inheritdoc/>
public class SDLWriter : SDLWriter<SDLWriter.DefaultWriteContext>
{
    /// <inheritdoc cref="IWriteContext"/>
    public class DefaultWriteContext : IWriteContext
    {
        /// <inheritdoc/>
        public TextWriter Writer { get; set; } = null!;
        /// <inheritdoc/>
        public Stack<ASTNode> Parents { get; set; } = new();
        /// <inheritdoc/>
        public int IndentLevel { get; set; } = 0;
        /// <inheritdoc/>
        public CancellationToken CancellationToken { get; set; }
    }

    /// <inheritdoc/>
    public SDLWriter() : base()
    {
    }

    /// <inheritdoc/>
    public SDLWriter(SDLWriterOptions options) : base(options)
    {
    }

    /// <inheritdoc cref="SDLWriter{TContext}"/>
    public virtual ValueTask PrintAsync(ASTNode? node, TextWriter writer, CancellationToken cancellationToken = default)
    {
        var context = new DefaultWriteContext()
        {
            Writer = writer,
            CancellationToken = cancellationToken,
        };
        return Visit(node, context);
    }
}
```

**Comment by Shane32** [comment]
Usage:
```cs
var result = await new SDLPrinter().PrintAsync(node, writer, token);
```

No need for extension method

**Comment by sungam3r** [comment]
I don't understand what to do with methods from interface.

**Comment by Shane32** [comment]
> I don't understand what to do with methods from interface.

I offered so many suggestions that I'm not sure which suggestion this comment applies to.  The most recent suggestion should work without any changes.

**Comment by sungam3r** [comment]
изображение

**Comment by Shane32** [comment]
I tried myself.  It would be a bit of a pain.  You'd have to create protected methods and also directly implement the interface, having the interface methods call the protected members.

The real problem is the OOP design here.  The `DefaultNodeVisitor` is serving two distinct purposes.  One is to walk the nodes and execute the proper methods.  The other is to provide a basic do-nothing implementation of an `INodeVisitor`.  For example, think of `IDocumentListener`.  The `DocumentListenerBase` is separate from the `DocumentExecuter`.  The executer walks the nodes while the `DocumentListenerBase` just provides an empty implementation of an `IDocumentListener`.

**Comment by sungam3r** [comment]
> The real problem is the OOP design here. 

😕  Frankly, I do not see any problems here. This is a well-known classic pattern and there is nothing to fix around. Classes and interface are clear and work well.

**Comment by Shane32** [comment]
Regardless of the design, my comments here still stand.  `ToSDL` should not exist as an extension method.  If it does exist, it should have an `Async` suffix.

**Comment by Shane32** [comment]
> 😕 Frankly, I do not see any problems here. This is a well-known classic pattern and there is nothing to fix around. Classes and interface are clear and work well.

I would disagree.

Here is a well known visitor pattern:

https://docs.microsoft.com/en-us/dotnet/api/system.linq.expressions.expressionvisitor?view=net-6.0

It does not implement an interface and the various Visit methods are protected, not public.  This is the pattern I would suggest we copy.

Do you have an example of a well known visitor pattern that you are copying?

**Comment by sungam3r** [comment]
> Here is a well known visitor pattern:

Implementation, not pattern. Our implementations differ in base interface. Actually I **can** remove interface and do all methods protected, not a big deal.

> Do you have an example of a well known visitor pattern that you are copying?

I did not copy anything. Here is one of the first links that I found now - https://thesharperdev.com/c-design-patterns-the-visitor-pattern/  . See `IExpressionVisitor` interface.

**Comment by sungam3r** [comment]
done

**Review by Shane32 — APPROVED** [review]
Looks great!!

**Pull request merged**